Updated doc local_development.md: new item in troublshooting section and detailed outline (#10380)
sourcegraph/sourcegraph: doc/dev/local_development.md

  - [Debug TypeScript code](#debug-typescript-code)
  - [Debug Go code](#debug-go-code)
- [Go dependency management](#go-dependency-management)
- [Codegen](#codegen)
- [Windows support](#windows-support)
- [Other nice things](#other-nice-things)
  - [Offline development](#offline-development)

## Environment

Sourcegraph server is a collection of smaller binaries. The development server, [dev/start.sh](https://github.com/sourcegraph/sourcegraph/blob/master/dev/start.sh), initializes the environment and starts a process manager that runs all of the binaries. See the [Architecture doc](architecture/index.md) for a full description of what each of these services does. The sections below describe the dependencies you need to run `dev/start.sh`.

<!-- omit in toc -->
### For Sourcegraph employees

[dev-private](https://github.com/sourcegraph/dev-private) repository has convenient preconfigured settings and external services on an enterprise account. You'll need to clone it to the same directory that contains this repository. After the initial setup you can run `enterprise/dev/start.sh` instead of `dev/start.sh`.

## Step 1: Install dependencies

> Sourcegraph has the following dependencies:
> > install instructions: [macOS](#macos), [Ubuntu](#ubuntu)

Reviewer: :D Sorry I wasn't clear, I mostly worried about the document formatting instead of the actual content here. It is weird to see a block quote inside the "list header" and its "list items".
```suggestion
> NOTE: Please see install instructions for [macOS](#macos) and [Ubuntu](#ubuntu) in succeeding sections.

Sourcegraph has the following dependencies:
```

> ## Outline

Author: 1. That would help me to build a mental map first. I actually do prefer detailed outlines especially when I come back to the document
2. It was autogenerated for me when I saved in VSCode. I suppose it was done by auto formatter :)
3.  I actually agree with that assessment. I thought the setup was a bit lengthy and manual (but not unmanageable at the same time).

> ## Outline

Reviewer: My rough observations:
1. https://docs.sourcegraph.com has TOC auto-generated for every page, but the fact is that me personally read dev docs more often in GitHub.com than in https://docs.sourcegrpah.com.
2. IIRC This is the only document we have TOC in the body of content.

> - [Getting started with developing Sourcegraph](#getting-started-with-developing-sourcegraph)

Reviewer: While the detailed the outline is very nice to have, is there a way to exclude this line? Seems  unreasonable to link back to the top of the document after read the document this far? WDYT?

Author: To be honest this outline was done automatically by prettier not handwritten.
So I think it should be either automated or not detailed

Reviewer 2: Let's delete it!

Author: delete detailed outline?

Reviewer 3: I would be surprised if Prettier generated outlines. Are you sure you don't have an extension installed that does this specifically?

Author: https://github.com/yzhang-gh/vscode-markdown

Author: ```
<!-- omit in toc -->
```
does the trick.

Reviewer: Nice trick!

      - [Increase maximum available file descriptors.](#increase-maximum-available-file-descriptors)
      - [Caddy 2 certificate problems](#caddy-2-certificate-problems)
      - [Running out of disk space](#running-out-of-disk-space)
  - [How to Run Tests](#how-to-run-tests)
  - [CPU/RAM/bandwidth/battery usage](#cpurambandwidthbattery-usage)
  - [How to debug live code](#how-to-debug-live-code)
    - [Debug TypeScript code](#debug-typescript-code)
    - [Debug Go code](#debug-go-code)
  - [Go dependency management](#go-dependency-management)
  - [Codegen](#codegen)
  - [Windows support](#windows-support)
  - [Other nice things](#other-nice-things)
    - [Offline development](#offline-development)

## Environment

Sourcegraph server is a collection of smaller binaries. The development server, [dev/start.sh](https://github.com/sourcegraph/sourcegraph/blob/master/dev/start.sh), initializes the environment and starts a process manager that runs all of the binaries. See the [Architecture doc](architecture/index.md) for a full description of what each of these services does. The sections below describe the dependencies you need to run `dev/start.sh`.

> ### For Sourcegraph employees

Reviewer: ```suggestion
### For Sourcegraph employees

```

> [dev-private](https://github.com/sourcegraph/dev-private) repo has convenient preconfigured repositories on an enterprise account. You'll need to clone it to the same directory that contains this repository. After the initial setup you can run `enterprise/dev/start.sh` instead of `dev/start.sh`.

Reviewer: ```suggestion
[dev-private](https://github.com/sourcegraph/dev-private) repository has convenient preconfigured settings and external services on an enterprise account. You'll need to clone it to the same directory that contains this repository. After the initial setup you can run `enterprise/dev/start.sh` instead of `dev/start.sh`.
```

## Step 1: Install dependencies

Sourcegraph has the following dependencies:

> > for install instruction see the section below

Reviewer: 🤔 Can you explain the intention of this line?

Reviewer 2: Agree this doesn't seem like it should be necessary. A README doesn't feel like the right place to document minimum version that we support anyway (this is prone to being out of date).

Author: I spent 1 hour installing dependencies from these links because I haven't read what is below. So this redundant line would have saved several hours for me.

./dev/drop-entire-local-database-and-redis.sh
```

#### Caddy 2 certificate problems

We use Caddy 2 to setup HTTPS for local development. It creates self-signed certificates and uses that to serve the local Sourcegraph instance. If your browser complains about the certificate, check the following:

1. The first time that Caddy 2 reverse-proxies your Sourcegraph instance, it needs to add its certificate authority to your local certificate store. This may require elevated permissions on your machine. If you haven't done so already, try running `caddy reverse-proxy --to localhost:3080` and enter your password if prompted. You may also need to run that command as the `root` user.

1. If you have completed the previous step and your browser still complains about the certificate, try restarting your browser or your local machine.

> #### Running out of disk space

Reviewer: ```suggestion
#### Running out of disk space

```

> if you see errors similar to this:

Reviewer: ```suggestion
If you see errors similar to this:

```

```
gitserver | ERROR cleanup: error freeing up space, error: only freed 1124101958 bytes, wanted to free 29905298227
```

> you are probably low on disk space. By default it tries to cleanup when there is less than 10% of available disk space.

Reviewer: ```suggestion

You are probably low on disk space. By default it tries to cleanup when there is less than 10% of available disk space.
```

> You can override that by setting this env variable

Reviewer: ```suggestion
You can override that by setting this env variable:

```

```bash
# means 5%. You may want to put that into .bashrc for convinience
SRC_REPOS_DESIRED_PERCENT_FREE=5
```

> 

Reviewer: ```suggestion
```

3. Go to `sourcegraph/sourcegraph`’s root directory and run `rm -rf node_modules`
3. Re-run the launch script (`./dev/start.sh`)

> 

Reviewer: ```suggestion
```
